src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,382 @@
+    /// 1) It allows access to the value by ref.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer(assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        const int DefaultPrimeSize = 3;
+        // 1-based index into _entries; 0 means empty
+        private int[] _buckets;
+        private Entry[] _entries;
+        // 0-based index into _entries of head of free chain: -1 means empty
+        private int _freeList = -1;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            // 0-based index of next entry in chain: -1 means empty
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = new int[DefaultPrimeSize];
+            _entries = new Entry[DefaultPrimeSize];
+        }
+
+        public DictionarySlim(int capacity)
+        {
+            capacity = HashHelpers.GetPrime(capacity);
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        // Drop sign bit to ensure non negative index
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private uint GetHashCode(TKey key) => (uint)key.GetHashCode();//key.GetHashCode() & 0x7FFFFFFF;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(int bucketIndex) => _buckets[bucketIndex] - 1;
+
+        public int Count { get; private set; }
+
+        public int Capacity { get => _entries.Length; }
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public bool Remove(TKey key)
```

Comment:
@danmosemsft What do you think? I'm tempted to take the array.copy cost in the remove and make everything else simpler. Idea being that a value can be defaulted and this dictionary is more for fast aggregation.

Reply:
@TylerBrinkley are you suggesting to shift the entire entries array along one? That makes remove into an O(n) operation whereas the free list is roughly O(1). If we are going to call this DictionarySlim users will expect it should have the same complexity characteristics as Dictionary. 

You almost have to do it in OrderedDictionary because you're relying on the entries array to provide the order -- as you noted you're OK with O(n) `Yes, O(1) lookups are very important to me. From my experience, collections are initialized and aren't modified after that point so mutation performance isn't as important to me. ` (Hopefully we will get community feedback on this approach.)

Reply:
Sorry, no. I wasn't suggesting making this an ordered dictionary. I just meant the way I was handling the remove method in my OrderedDictionary class seemed like it might be more performant but really I have no idea.

Reply:
Ah right. Yeah, whatever approach we use, we should retain the complexity characteristics of Dictionary. One of which is, I can remove lots of elements in a loop from a huge dictionary, and my app will not just hang 😮

Reply:
The gaps can be filled in easily enough without shifting the entire collection - just have to swap the last element in the list into the newly opened slot. Without back pointers, I think it would still be necessary to hash the moved element and follow its chain to repair the break, but it would at least be an O(logn) operation rather than O(n)

Reply:
@Zhentar good idea. Wouldn't it be O(1) not O(logn) ? The chains should be of order unity if we have a reasonably distributed hash code.
If anyone wants to try this, they should measure perf in various scenarios - the current system at least has the advantage that Dictionary has proved it should not be bad so I think it is fine to leave it for now.
I suppose that moving entries around would decrease the locality in the entries array that entries that were added around the same time currently achieve. Maybe that only matters in artificial scenarios.

Reply:
Actually locality might improve in some scenarios, since it would mean all new adds always go at the end of Entries, even if there were previous removes. It would need measurements/scenarios to know.

Reply:
If implemented, you'd be able to remove the _freeList field as well as the _count fields in the enumerators and wouldn't have to do so many checks in those enumerators' `MoveNext` methods. I think it's worth pursuing.

Reply:
@Zhantar @TylerBrinkley It works well, thanks. Remove is still faster than Dictionary.